File: app/src/main/java/com/ad/whatsappclone/activity/SignUpActivity.java
package com.ad.whatsappclone.activity;

import android.app.ProgressDialog;
import android.content.Intent;
import android.os.Bundle;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.ad.whatsappclone.databinding.ActivitySignUpBinding;
import com.ad.whatsappclone.models.Constraints;
import com.ad.whatsappclone.models.Users;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.FirebaseDatabase;

public class SignUpActivity extends AppCompatActivity {
    ActivitySignUpBinding binding;
    private FirebaseAuth mAuth;
    FirebaseDatabase database;
    ProgressDialog progressDialog;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivitySignUpBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        if (getSupportActionBar() != null) {
            getSupportActionBar().hide();
        }
        mAuth = FirebaseAuth.getInstance();
        database = FirebaseDatabase.getInstance();


        progressDialog = new ProgressDialog(SignUpActivity.this);
        progressDialog.setTitle("Creating Account");
        progressDialog.setMessage("We are creating your account");

        binding.signupBtnSignup.setOnClickListener(view -> {

            progressDialog.show();

            mAuth.createUserWithEmailAndPassword(binding.signupEdtEmail.getText().toString(),
                    binding.signupEdtPassword.getText().toString())
                    .addOnCompleteListener(task -> {
                        progressDialog.dismiss();
                        if (task.isSuccessful()) {
                            Users user = new Users(binding.signupEdtUsername.getText().toString(),
                                    binding.signupEdtPassword.getText().toString(),
                                    binding.signupEdtEmail.getText().toString());

                            String userId = task.getResult().getUser().getUid();
                            database.getReference().child(Constraints.USER_NODE).child(userId).setValue(user);

                            Toast.makeText(SignUpActivity.this, "User Created successfully", Toast.LENGTH_SHORT).show();
                        } else {
                            Toast.makeText(SignUpActivity.this, task.getException().getMessage(), Toast.LENGTH_SHORT).show();
                        }
                    });
        });

        binding.signupTxtAlreadyHaveAccount.setOnClickListener(view -> startActivity(new Intent(SignUpActivity.this, SignInActivity.class)));
    }
}

File: app/src/main/java/com/ad/whatsappclone/ui/authentication/SignUpActivity.kt
package com.ad.whatsappclone.ui.authentication

import android.app.ProgressDialog
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Toast
import com.ad.whatsappclone.databinding.ActivitySignUpBinding
import com.ad.whatsappclone.models.Constraints
import com.ad.whatsappclone.models.Users
import com.ad.whatsappclone.ui.BaseActivity
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.FirebaseDatabase
import java.util.*

class SignUpActivity : BaseActivity() {
    private lateinit var binding: ActivitySignUpBinding
    private var mAuth: FirebaseAuth? = null
    var database: FirebaseDatabase? = null
    var progressDialog: ProgressDialog? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivitySignUpBinding.inflate(layoutInflater)
        setContentView(binding.root)
        hideActionBar()
        mAuth = FirebaseAuth.getInstance()
        database = FirebaseDatabase.getInstance()
        progressDialog = ProgressDialog(this@SignUpActivity)
        progressDialog!!.setTitle("Creating Account")
        progressDialog!!.setMessage("We are creating your account")
        binding.signupBtnSignup.setOnClickListener { view: View? ->
            progressDialog!!.show()
            mAuth!!.createUserWithEmailAndPassword(
                binding.signupEdtEmail.text.toString(),
                binding.signupEdtPassword.text.toString()
            )
                .addOnCompleteListener { task: Task<AuthResult> ->
                    progressDialog!!.dismiss()
                    if (task.isSuccessful) {
                        val user = Users(
                            binding.signupEdtUsername.text.toString(),
                            binding.signupEdtPassword.text.toString(),
                            binding.signupEdtEmail.text.toString()
                        )
                        val userId = Objects.requireNonNull(task.result.user)?.uid
                        if (userId != null) {
                            database!!.reference.child(Constraints.USER_NODE).child(userId)
                                .setValue(user)
                        }

                        Toast.makeText(
                            this@SignUpActivity,
                            "User Created successfully",
                            Toast.LENGTH_SHORT
                        ).show()
                    } else {
                        Toast.makeText(
                            this@SignUpActivity,
                            task.exception!!.message,
                            Toast.LENGTH_SHORT
                        ).show()
                    }
                }
        }
        binding.signupTxtAlreadyHaveAccount.setOnClickListener {
            startActivity(
                Intent(this@SignUpActivity, SignInActivity::class.java)
            )
        }
    }
}